## 3. Containerising a WGS workflow

 [\<\< Episode 2](https://github.com/PawseySC/bio-workshop-18/blob/master/2.fastqc.md)
 | [TOC](https://github.com/PawseySC/bio-workshop-18/blob/master/README.md) |
 [Episode 4 \>\>](https://github.com/PawseySC/bio-workshop-18/blob/master/4.hpc.md)
______

### Before we start
Change directory to the path relevant for this Episode:

    cd bio-workshop-18/episode3_wgs_workflow/


---
### A sample bioinformatics workflow
The `data_files/` folder contains two paired read files from a plant called *Fagopyrum tartaricum*, `SRR6166481_sub_1.fastq.gz` and `SRR6166481_sub_2.fastq.gz`, as well as a reference genome from another plant of the same Genus, namely *Fagopyrum_esculentum*, file `Fagopyrum_esculentum.fasta`. To allow for the workflow to be run in a short timeframe, the number of reads has been limited to 100,000.

From inside `episode3_wgs_workflow/`, we are going to work with the following sample workflow:

![Workflow diagram - Episode 3](episode3_wgs_workflow/Workflow.png)

This is meant to be a toy workflow, to illustrate how different types of packages can be containerised with Docker, and also to comment on some specific issues that can arise with certain containers.


---
### Hands-on: let us build a containerised workflow
We will build the workflow above using the skills acquired in [Episode 1](https://github.com/PawseySC/bio-workshop-18/blob/master/1.containers.md) and [Episode 2](https://github.com/PawseySC/bio-workshop-18/blob/master/2.fastqc.md).

The scope of this Episode is not to build the workflow from scratch, but rather to port it from a preexisting, non-containerised version. The full sequence of commands corresponding to the latter are written inside a file named `workflow_start_here.sh`.   required
Also, this original scriptfile has been cut into six bash scripts, whose names start with numbers `1` to `6`; each of these scripts corresponds to a step in the workflow above.

Ideally, we would search for each required package in [QUAY Hub](https://quay.io) or [Docker Hub](https://hub.docker.com). However, to save time during the workshop, the list of repositories to be downloaded is available in the file named `list_of_containers`.

Here are our goals:
1. write a script to pull all the container repositories required for the workflow;
2. for each step in the workflow, adapt the provided bash script so that it runs the bioinformatics packages using Docker containers;
3. optionally, write a driver bash script that executes all the step scripts in sequence.


---
### Tips and hints
Some useful tips:
- you will need to specify some Docker flags, as in [Episode 2](https://github.com/PawseySC/bio-workshop-18/blob/master/2.fastqc.md); an environment variable can again be useful for the purpose;
- for the pull and the driver scripts, you can just write a sequence of commands, or ... use a `for` loop.

Some hints for special cases:
- for Java packages, you will need to inspect the container to find out the location of the .jar file;
- for Trimmomatic, adapters are in the `adapters/` directory inside the Episode directory;
- output redirection with `>` works straight away with Docker containers;
- input redirection with `<` and piping with `|` requires some care, with an extra Docker flag required (review those discussed in [Episode 1](https://github.com/PawseySC/bio-workshop-18/blob/master/1.containers.md)).


---
### Note on Trimmomatic: a Java container (step 2)
In the cleaning step, to run **Trimmomatic** you need a command like:

    java -jar ${PATH_TO_TRIMMOMATIC_JAR}/trimmomatic.jar

Where is `trimmomatic.jar` located inside the container? Let us find out by means of an interactive container session:

    docker run -it --rm quay.io/biocontainers/trimmomatic:0.38--0 bash

Inside the container, let us look for the jar file and then exit:

    find / -name "*trimmomatic*.jar" && exit

The output will solve your quest:

    /usr/local/share/trimmomatic-0.38-0/trimmomatic.jar

As mentioned in [Episode 2](https://github.com/PawseySC/bio-workshop-18/blob/master/2.fastqc.md), biocontainers in QUAY Hub have usually packages installed into `/usr/local/`.

Finally, some further inspection of this **Trimmomatic** container reveals there is a `trimmomatic` executable script available. This script can be used in substitution of the typical way of invoking Java packages, so that both of these will work:

    docker run <options> <cont ID> java -jar /usr/local/share/trimmomatic-0.38-0/trimmomatic.jar <arguments>
and

    docker run <options> <cont ID> trimmomatic <arguments>


---
### Note on piping with samtools and bcftools (step 6)
In the SNP calling step, two command lines in the starting workflow contain piping operations with `|`, something like:

    app1 <args1> | app2 <args2>

You can try and containerise this line as you did in the previous steps:

    run_flags="--rm -v $(pwd):/data -w /data"
    docker run $run_flags $cont1 app1 <args1> | docker run $run_flags $cont2 app2 <args2>

However, you will get an error similar to:

    write /dev/stdout: broken pipe

This happens because the first command tries to send output to the second one, but does not succeed because the second container does not have the standard input open.

How to open it? Well, just re-use a Docker flag we saw in [Episode 1](https://github.com/PawseySC/bio-workshop-18/blob/master/1.containers.md)) when talking about interactive sessions: `-i`:

    docker run $run_flags $cont1 app1 <args1> | docker run -i $run_flags $cont2 app2 <args2>

This will do the job!


---
### Complications when looking for containers in Web repositories

The list of repositories contained in the file "`containers`" has some useful behind the scenes. In particular:

- For **Mummer**, one could try and use the containers for **Version 3.x** from both QUAY Hub and Docker Hub; however, one has a broken `mummer` script, and the other one has a broken `mummerplot` script.  
The container for **Mummer4** from QUAY Hub works fine and is the one used in this Episode.

- For **bcftools**, the container from QUAY Hub has a broken `vcfutils.pl`, as it comes with no Perl installed!  
The container from Docker Hub works fine and has been used in this Episode.

These types of issues do not happen often, but can still happen.  
What if no working container can be found in the Web repositories? Well, you can consider build your own, by writing a Dockerfile: see [6](https://github.com/PawseySC/bio-workshop-18/blob/master/6.cellranger.md).



______
 [\<\< Episode 2](https://github.com/PawseySC/bio-workshop-18/blob/master/2.fastqc.md)
 | [TOC](https://github.com/PawseySC/bio-workshop-18/blob/master/README.md) |
 [Episode 4 \>\>](https://github.com/PawseySC/bio-workshop-18/blob/master/4.hpc.md)
