## 3. Containerising a WGS workflow

 [\<\< Episode 2](https://github.com/PawseySC/bio-workshop-18/blob/master/2.fastqc.md)
 | [TOC](https://github.com/PawseySC/bio-workshop-18/blob/master/README.md) |
 [Episode 4 \>\>](https://github.com/PawseySC/bio-workshop-18/blob/master/4.hpc.md)
______

### Before we start
Change directory to the path relevant for this Episode:

    cd bio-workshop-18/episode3_wgs_workflow/


---
### A sample bioinformatics workflow
The folder `bio-workshop-18/data_files/` contains two paired read files from a plant called *Fagopyrum tartaricum*, `SRR6166481_sub_1.fastq.gz` and `SRR6166481_sub_2.fastq.gz`, as well as a reference genome from another plant of the same Genus, namely *Fagopyrum_esculentum*, file `Fagopyrum_esculentum.fasta`. To allow for the workflow to be run in a short timeframe, the number of reads has been limited to 100,000.

From inside `episode3_wgs_workflow/`, we are going to work with the following sample workflow:

![Workflow diagram - Episode 3](episode3_wgs_workflow/Workflow.png)

This is meant to be a toy workflow, to illustrate how different types of packages can be containerised with Docker, and also to comment on some specific issues that can arise with certain containers.


---
### Hands-on: let us build a containerised workflow
We will build the workflow above using the skills acquired in [Episode 1](https://github.com/PawseySC/bio-workshop-18/blob/master/1.containers.md) and [Episode 2](https://github.com/PawseySC/bio-workshop-18/blob/master/2.fastqc.md).

The scope of this Episode is not to build the workflow from scratch, but rather to port it from a preexisting, non-containerised version. The full sequence of commands corresponding to the latter are written inside a file named `workflow_start_here.sh`. We can inspect the content of this script with `cat`:

    cat workflow_start_here.sh

For convenience this original scriptfile has been cut into six bash scripts, whose names start with numbers `1` to `6`; each of these scripts corresponds to a step in the workflow above:

```
ls -1 *.sh

1.fastqc.sh
2.trimmomatic.sh
3.bowtie2.sam.sh
4.velvet.sh
5.mummer4.sh
6.bcftools.sam.sh
workflow_start_here.sh
```

Ideally, we would search for each required package in [QUAY Hub](https://quay.io) or [Docker Hub](https://hub.docker.com). However, to save time during the workshop, the list of repositories to be downloaded is available in the file named `list_of_containers`:

```
cat list_of_containers

quay.io/biocontainers/fastqc:0.11.7--4
quay.io/biocontainers/trimmomatic:0.38--0
quay.io/biocontainers/bowtie2:2.3.4.2--py27h2d50403_0
quay.io/biocontainers/samtools:1.9--h46bd0b3_0
quay.io/biocontainers/velvet:1.2.10--1
quay.io/biocontainers/mummer4:4.0.0beta2--pl526hfc679d8_3
biocontainers/bcftools:v1.3.1-1b1-deb_cv1
```

Here are our goals:
1. write a script for a one-off pull of all container repositories required for the workflow;
2. for each step in the workflow, edit the provided bash script so that it runs the bioinformatics packages using Docker containers;
3. optionally, write a driver bash script that executes all the step scripts in sequence.


---
### Pull the required packages
We need to write a script that runs a set of `docker pull` to download the required repositories.  
To save on typing, let us start from the file  `list_of_containers`:

    cp list_of_containers pull_containers.sh
    
    nano pull_containers.sh

We are using `nano`, a text editor, for editing the new file `pull_containers.sh` to prepend the command `docker pull` to each repository name. We will also need to add the bash shebang `#!/bin/bash` at the beginning of the script. The final result is:

```
#!/bin/bash

docker pull quay.io/biocontainers/fastqc:0.11.7--4
docker pull quay.io/biocontainers/trimmomatic:0.38--0
docker pull quay.io/biocontainers/bowtie2:2.3.4.2--py27h2d50403_0
docker pull quay.io/biocontainers/samtools:1.9--h46bd0b3_0
docker pull quay.io/biocontainers/velvet:1.2.10--1
docker pull quay.io/biocontainers/mummer4:4.0.0beta2--pl526hfc679d8_3
docker pull biocontainers/bcftools:v1.3.1-1b1-deb_cv1
```

Let us save the file in `nano` using `Ctrl-O` (`Enter` to confirm) and then quit with `Ctrl-X`.

Now, let us grant execution permission to the script we created:

    chmod +x pull_containers.sh

And finally run it:

    ./pull_containers.sh

After a few minutes of processing, the container images required for our workflow have been downloaded.


---
### 1. Quality control with FastQC

As this step involves exactly the same command discussed in [Episode 2](https://github.com/PawseySC/bio-workshop-18/blob/master/2.fastqc.md), the script `1.fastqc.sh` comes ready to be run. Let us have a look at its contents:

```
cat 1.fastqc.sh


#!/bin/bash
run_flags="--rm -v $(pwd):/data -w /data"

cp -p ../data_files/SRR6166481_sub_[12].fastq.gz .

docker run $run_flags quay.io/biocontainers/fastqc:0.11.7--4 fastqc -o . SRR6166481_sub_1.fastq.gz

docker run $run_flags quay.io/biocontainers/fastqc:0.11.7--4 fastqc -o . SRR6166481_sub_2.fastq.gz
```

Because we are going to use the same flags for all `docker run` commands in all these scripts, we are using a variable called `run_flags` to store their values and then use them when required. This is simpler to type and less error prone:

    run_flags="--rm -v $(pwd):/data -w /data"

Now let us run this step:

    ./1.fastqc.sh


---
### 2. Cleaning with Trimmomatic - a Java container
Let us edit the script with `nano 2.trimmomatic.sh` (see instructions in the section **Pull the required packages** above).

First, note the use of "`\`" at end of lines to break a long line into multiple ones and improve readability.

We need to prepend

    docker run $run_flags quay.io/biocontainers/trimmomatic:0.38--0 

to the `java` execution line.

Now let us read through the `java` arguments to see if we need to provide further information. The command is of the form

    java -jar <path-to-trimmomatic-jar>/trimmomatic.jar

Where is `trimmomatic.jar` located inside the container? Let us find out. We need to temporarily close `nano` and start an interactive container session:

    docker run -it --rm quay.io/biocontainers/trimmomatic:0.38--0 bash

Inside the container, let us look for the jar file and then exit:

    find / -name "*trimmomatic*.jar" && exit

The output will solve your quest:

    /usr/local/share/trimmomatic-0.38-0/trimmomatic.jar

As mentioned in [Episode 2](https://github.com/PawseySC/bio-workshop-18/blob/master/2.fastqc.md), biocontainers in QUAY Hub have usually packages installed into `/usr/local/`. We can now `exit` the interactive session:

    exit

Now we can complete the path to `trimmomatic.jar` in the script file using `nano`.

There is another file required by **Trimmomatic**, the one containing the adapter sequences. In this example the command line will look for `adapters/all_PE.fa`, which for convenience is already provided in the Episode directory.

We can now run the script for step 2:

    ./2.trimmomatic.sh

**Advanced note:** some further inspection of this **Trimmomatic** container reveals there is a `trimmomatic` executable script available. This script can be used in substitution of the typical way of invoking Java packages, so that both of these will work:

    docker run <options> <cont ID> java -jar /usr/local/share/trimmomatic-0.38-0/trimmomatic.jar <arguments>
and

    docker run <options> <cont ID> trimmomatic <arguments>


---
### 3. Read alignment with Bowtie2 and samtools
Let us edit



---
### Tips and hints
Some useful tips:
- you will need to specify some Docker flags, as in [Episode 2](https://github.com/PawseySC/bio-workshop-18/blob/master/2.fastqc.md); an environment variable can again be useful for the purpose;

Some hints for special cases:
- output redirection with `>` works straight away with Docker containers;
- input redirection with `<` and piping with `|` requires some care, with an extra Docker flag required (review those discussed in [Episode 1](https://github.com/PawseySC/bio-workshop-18/blob/master/1.containers.md)).


---
### Note on piping with samtools and bcftools (step 6)
In the SNP calling step, two command lines in the starting workflow contain piping operations with `|`, something like:

    app1 <args1> | app2 <args2>

You can try and containerise this line as you did in the previous steps:

    run_flags="--rm -v $(pwd):/data -w /data"
    docker run $run_flags $cont1 app1 <args1> | docker run $run_flags $cont2 app2 <args2>

However, you will get an error similar to:

    write /dev/stdout: broken pipe

This happens because the first command tries to send output to the second one, but does not succeed because the second container does not have the standard input open.

How to open it? Well, just re-use a Docker flag we saw in [Episode 1](https://github.com/PawseySC/bio-workshop-18/blob/master/1.containers.md)) when talking about interactive sessions: `-i`:

    docker run $run_flags $cont1 app1 <args1> | docker run -i $run_flags $cont2 app2 <args2>

This will do the job!


---
### Complications when looking for containers in Web repositories

The list of repositories contained in the file "`containers`" has some useful behind the scenes. In particular:

- For **Mummer**, one could try and use the containers for **Version 3.x** from both QUAY Hub and Docker Hub; however, one has a broken `mummer` script, and the other one has a broken `mummerplot` script.  
The container for **Mummer4** from QUAY Hub works fine and is the one used in this Episode.

- For **bcftools**, the container from QUAY Hub has a broken `vcfutils.pl`, as it comes with no Perl installed!  
The container from Docker Hub works fine and has been used in this Episode.

These types of issues do not happen often, but can still happen.  
What if no working container can be found in the Web repositories? Well, you can consider build your own, by writing a Dockerfile: see [6](https://github.com/PawseySC/bio-workshop-18/blob/master/6.cellranger.md).



______
 [\<\< Episode 2](https://github.com/PawseySC/bio-workshop-18/blob/master/2.fastqc.md)
 | [TOC](https://github.com/PawseySC/bio-workshop-18/blob/master/README.md) |
 [Episode 4 \>\>](https://github.com/PawseySC/bio-workshop-18/blob/master/4.hpc.md)
