## Objective ##
Run a real-world bioinformatics application in a docker container

### Running BLAST from a container ###
We'll be running a BLAST (Basic Local Alignment Search Tool) example with a container from [BioContainers](https://biocontainers.pro).

To begin, we'll move to the directory:

```
cd bio-workshop-18/episode5_blast
```

There is a script, `blast-demo.sh`, that will run all the commands if you want.  Otherwise, you
can follow along below.

To start we'll pull the BLAST container (this will take a little bit):

```
docker pull biocontainers/blast
Using default tag: latest
latest: Pulling from biocontainers/blast
22dc81ace0ea: Pull complete
1a8b3c87dba3: Pull complete
...
Status: Downloaded newer image for biocontainers/blast:latest
```

We can run a simple command to verify the container works:

```
docker run biocontainers/blast blastp -help

USAGE
  blastp [-h] [-help] [-import_search_strategy filename]
...
 -use_sw_tback
   Compute locally optimal Smith-Waterman alignments?
```

Let's copy some data over to start working with:

```
cp -p ../data_files/P04156.fasta .
cp -p ../data_files/zebrafish.1.protein.faa.gz .
gunzip zebrafish.1.protein.faa.gz
```

This is a human prion FASTA sequence along with a reference database we'll blast against.  We need to prepare the zebrafish database with `makeblastdb`:

```
docker run --rm -v $(pwd):/data/ -w /data biocontainers/blast makeblastdb -in zebrafish.1.protein.faa -dbtype prot
```

We can now run BLAST and compare our data against the database:

```
docker run --rm -v $(pwd):/data/ -w /data biocontainers/blast blastp -query P04156.fasta -db zebrafish.1.protein.faa -out results.txt
```

The final results are stored in `results.txt`:

```
> less results.txt
                                                                     Score     E
equences producing significant alignments:                          (Bits)  Value

 XP_017207509.1 protein piccolo isoform X2 [Danio rerio]             43.9    2e-04
 XP_017207511.1 mucin-16 isoform X4 [Danio rerio]                    43.9    2e-04
 XP_021323434.1 protein piccolo isoform X5 [Danio rerio]             43.5    3e-04
 XP_017207510.1 protein piccolo isoform X3 [Danio rerio]             43.5    3e-04
 XP_021323433.1 protein piccolo isoform X1 [Danio rerio]             43.5    3e-04
 XP_009291733.1 protein piccolo isoform X1 [Danio rerio]             43.5    3e-04
 NP_001268391.1 chromodomain-helicase-DNA-binding protein 2 [Dan...  35.8    0.072
...
```


### Conclusion ###
There are a lot of applications (not just bioinformatics) already wrapped up in container images.  Here's a small list of some of the repositories we use at Pawsey:

* [DockerHub](hub.docker.com)
* [Bioboxes](bioboxes.org)
* [Biocontainers](biocontainers.pro)
* [Nvidia GPU Cloud (NGC)](ngc.nvidia.com)*
* [quay.io](quay.io)*

The last two require you to create an account and login to access containers.

### Best practices ###

- don't re-invent the wheel, it's worth looking to see who's done what you want already
